java-checks-test-sources/src/main/java/checks/security/UnpredictableSaltCheck.java

```diff
@@ -0,0 +1,107 @@
+package checks.security;
+
+import java.nio.charset.StandardCharsets;
+import java.security.MessageDigest;
+import java.security.NoSuchAlgorithmException;
+import javax.crypto.spec.PBEKeySpec;
+
+public class UnpredictableSaltCheck {
+
+  public void testDigestWithParamAndOneUpdate(byte[] salt, String passwordToHash) throws NoSuchAlgorithmException {
+
+    MessageDigest md = MessageDigest.getInstance("SHA-512");
+    md.update(passwordToHash.getBytes(StandardCharsets.UTF_8));
+    byte[] hashedPassword = md.digest(salt); // Compliant
+
+  }
+
+  public void testDigestWithOneUpdateAndWithoutParams(String passwordToHash) throws NoSuchAlgorithmException {
+    MessageDigest md = MessageDigest.getInstance("SHA-512");
+    md.update(passwordToHash.getBytes(StandardCharsets.UTF_8));
+    byte[] hashedPassword = md.digest(); // Noncompliant [[sc=29;ec=40]] {{Add an unpredictable salt value to this hash.}}
+  }
+
+  public void testDigestWithMultipleUpdates(String passwordToHash) throws NoSuchAlgorithmException {
+    MessageDigest md = MessageDigest.getInstance("SHA-512");
+    md.update(passwordToHash.getBytes(StandardCharsets.UTF_8));
+    md.update(passwordToHash.getBytes(StandardCharsets.UTF_8));
+    md.update(passwordToHash.getBytes(StandardCharsets.UTF_8));
+    byte[] hashedPassword = md.digest(); // Compliant
+    
+    md.getAlgorithm();
+    
+    new MyMessageDigest("").update(passwordToHash.getBytes());
+  }
+
+  public void testDigestWithoutUpdate(String passwordToHash) throws NoSuchAlgorithmException {
+    MessageDigest md = MessageDigest.getInstance("SHA-512");
+    byte[] hashedPassword = md.digest(passwordToHash.getBytes(StandardCharsets.UTF_8)); // Noncompliant
+  }
+
+  public void testPBESpec(char[] chars, byte[] salt) throws NoSuchAlgorithmException {
+    PBEKeySpec spec = new PBEKeySpec(chars); // Noncompliant [[sc=23;ec=44]]
+    PBEKeySpec spec1 = new PBEKeySpec(chars, salt, 1); // Compliant 
+  }
+
+  public void testNonRandomPBESpec(char[] chars, String str) throws NoSuchAlgorithmException {
+    byte[] salt = "notrandom".getBytes();
+    PBEKeySpec spec = new PBEKeySpec(chars, salt, 1); // Noncompliant [[sc=23;ec=53]] {{Make this salt unpredictable.}}
```

Comment:
In the case the salt is predictable and comes from a variable, we might want to add a secondary location to clearly point why the salt is predictable.